### PR TITLE
Added support to view the Encoding history for MVS and USS in the History View

### PR DESCRIPTION
**Pull request by KaranP25**
## Release Notes

Milestone:

Changelog: Added support to view the Encoding history for MVS and Dataset in the History View.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ ] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer/src/shared/HistoryView.ts [hunk 1]
```diff
@@ -95,9 +97,14 @@ export class HistoryView extends WebView {
             fileHistory: treeProvider.getFileHistory(),
             dsTemplates: type === "ds" ? (treeProvider as DatasetTree).getDsTemplates() : undefined,
             favorites: treeProvider.getFavorites(),
+            encodingHistory: type === "uss" || type === "ds" ? this.fetchEncodingHistory() : [],
         };
     }
 
+    private fetchEncodingHistory(): string[] {
+        return ZoweLocalStorage.getValue<string[]>("zowe.encodingHistory") ?? [];
```
Comment (zFernand0) [inline packages/zowe-explorer/src/shared/HistoryView.ts:105]: I believe we should move the `"zowe.encodingHistory"` into the `LocalStorageKey` enum.

---

Also, thanks for doing `?? []` at the end.
This is very similar to how [this issue](#2770) below was fixed in [this PR](#2783)

File: packages/zowe-explorer/src/shared/HistoryView.ts [hunk 2]
```diff
@@ -144,6 +151,18 @@ export class HistoryView extends WebView {
                     });
                 }
                 break;
+            case "encodingHistory":
+                Object.keys(message.attrs.selectedItems).forEach((selectedItem) => {
+                    if (message.attrs.selectedItems[selectedItem]) {
+                        //(treeProvider as USSTree).removeEncodingHistory(selectedItem);
```
Comment (zFernand0) [inline packages/zowe-explorer/src/shared/HistoryView.ts:157]: Should we remove this commented out code?

File: packages/zowe-explorer/src/shared/HistoryView.ts [hunk 3]
```diff
@@ -168,6 +187,10 @@ export class HistoryView extends WebView {
                         treeProvider.resetFileHistory();
                     }
                     break;
+                case "encodingHistory":
+                    //(treeProvider as USSTree).resetEncodingHistory();
```
Comment (zFernand0) [inline packages/zowe-explorer/src/shared/HistoryView.ts:191]: same as previous comment

File: packages/zowe-explorer/src/shared/utils.ts [hunk 4]
```diff
@@ -473,7 +474,7 @@ export async function promptForEncoding(node: IZoweDatasetTreeNode | IZoweUSSTre
     } else if (node.encoding === null || currentEncoding === "text") {
         currentEncoding = ebcdicItem.label;
     }
-    const encodingHistory = ZoweLocalStorage.getValue<string[]>("zowe.encodingHistory") ?? [];
+    const encodingHistory = ZoweLocalStorage.getValue<string[]>("zowe.encodingHistory") ?? []; //TreeProviders.uss.getEncodingHistory() ?? [];
```
Comment (zFernand0) [inline packages/zowe-explorer/src/shared/utils.ts:477]: Once we move the `"zowe.encodingHistory"` to the enum mentioned in a previous comment, I think we can change it across the board (without causing a circular import)

Also, don't forget to remove any commented out code 😋

File: packages/zowe-explorer/src/webviews/src/edit-history/components/PersistentToolBar/PersistentClearAllButton.tsx [hunk 5]
```diff
@@ -33,7 +33,7 @@ export default function PersistentClearAllButton(): JSXInternal.Element {
   const renderClearAllButton = () => {
     const clearAllText = localize("PersistentClearAllButton.clearAll", "Clear All");
 
-    return selection[type] === "search" || selection[type] === "fileHistory" ? (
+    return selection[type] === "search" || selection[type] === "fileHistory" || selection[type] === "encodingHistory" ? (
```
Comment (zFernand0) [inline packages/zowe-explorer/src/webviews/src/edit-history/components/PersistentToolBar/PersistentClearAllButton.tsx:36]: Curious if we should make turn these items into a type (or something we can iterate over) and use it across the `edit-history` components

That way we may be able to simplify this to just  `selection[type] in EditHistory.SelectionType`

**Review by zFernand0 — CHANGES_REQUESTED** [review]
Changes LGTM! 😋 
Left a few small requests/comments 🥳

File: packages/zowe-explorer/src/shared/HistoryView.ts (hunk 1 above)
Comment (KaranP25) [reply packages/zowe-explorer/src/shared/HistoryView.ts:105]: Hi, Where is this enum located?

**Comment by rudyflores** [comment]
@zFernand0 I see the changelog check failing, even though the changelog seems to be in place in this PR. Do you know if this is a known issue for the check to break when changes are made specifically to our webview code?

File: packages/zowe-explorer/src/shared/HistoryView.ts (hunk 1 above)
Comment (zFernand0) [reply packages/zowe-explorer/src/shared/HistoryView.ts:105]: Oh, my apologies, I mixed up my PR reviews.
it was added only in the `next` branch in [this PR](https://github.com/zowe/zowe-explorer-vscode/pull/2788)
**_Source:_** [packages/zowe-explorer/src/utils/ZoweLocalStorage.ts](https://github.com/zowe/zowe-explorer-vscode/blob/f3cebbd8f00ad84173caa2ff67bb9e342f8fbc81/packages/zowe-explorer/src/utils/ZoweLocalStorage.ts)


No need to create a new enum just for these.

**Comment by zFernand0** [comment]
> zFernand0 I see the changelog check failing, even though the changelog seems to be in place in this PR. Do you know if this is a known issue for the check to break when changes are made specifically to our webview code?

I kind of agree that it probably doesn't need to be looking for that since it's [a private package](https://github.com/zowe/zowe-explorer-vscode/blob/main/packages/zowe-explorer/src/webviews/package.json#L3). We can let this PR just have that failing check, or add the `no-changelog` label so that it doesn't look for changelog-updates in new commits.

Ideally, we would enhance the checklog checker to have a flag/setting that doesn't look for changelog updates in private packages of a monorepo 😋

File: packages/zowe-explorer/src/shared/utils.ts [hunk 6]
```diff
@@ -507,7 +508,7 @@ export async function promptForEncoding(node: IZoweDatasetTreeNode | IZoweUSSTre
                 encoding = { kind: "other", codepage: response };
                 const filterEncodingList = encodingHistory.filter((codepage) => codepage !== response);
                 filterEncodingList.unshift(encoding.codepage);
-                ZoweLocalStorage.setValue("zowe.encodingHistory", filterEncodingList.slice(0, globals.MAX_FILE_HISTORY));
+                ZoweLocalStorage.setValue("zowe.encodingHistory", filterEncodingList.slice(0, globals.MAX_FILE_HISTORY)); //TreeProviders.uss.addEncodingHistory(encoding.codepage);
```
Comment (rudyflores) [inline packages/zowe-explorer/src/shared/utils.ts:511]: Remove commented code from here as well :)

File: packages/zowe-explorer/src/webviews/src/edit-history/components/PersistentTable/PersistentTableData.tsx [hunk 7]
```diff
@@ -35,7 +35,7 @@ export default function PersistentTableData({ persistentProp }: Readonly<{ persi
   };
 
   const renderSelectButton = (item: string, i: number) => {
-    return selection[type] === "search" || selection[type] === "fileHistory" ? (
+    return selection[type] === "search" || selection[type] === "fileHistory" || selection[type] === "encodingHistory" ? (
```
Comment (rudyflores) [inline packages/zowe-explorer/src/webviews/src/edit-history/components/PersistentTable/PersistentTableData.tsx:38]: If you go ahead and do the refactor suggested by @zFernand0 , might be good to add it here as well!

**Review by rudyflores — CHANGES_REQUESTED** [review]
Thanks for making this change, left some comments regarding some potential edits to the code changes

**Comment by KaranP25** [comment]
Thank you @zFernand0 and @rudyflores for your comments, I have now addressed them with my latest commit. Feel free to take a look and comment on anything else I might have missed.

**Review by JillieBeanSim — APPROVED** [review]
LGTM! thanks @KaranP25

File: packages/zowe-explorer/src/shared/utils.ts [hunk 8]
```diff
@@ -505,8 +505,8 @@ export async function promptForEncoding(node: IZoweDatasetTreeNode | IZoweUSSTre
             });
             if (response != null) {
                 encoding = { kind: "other", codepage: response };
-                const filterEncodingList = encodingHistory.filter((codepage) => codepage !== response);
-                filterEncodingList.unshift(encoding.codepage);
+                const filterEncodingList = encodingHistory.filter((codepage) => codepage.toUpperCase() !== response.toUpperCase());
+                filterEncodingList.unshift(encoding.codepage.toUpperCase());
```
Comment (zFernand0) [inline packages/zowe-explorer/src/shared/utils.ts:509]: really good catch 😋 
Do we want to add a small test for this?

**Review by zFernand0 — APPROVED** [review]
LGTM! 😋 
Functionality works great too! 🥳 

Random question:
Are we concerned that MVS and USS share the same encoding-history?

I don't think it has been a problem so far, but with this enhancement we are giving more visibility to this information. 

I'm just wondering if people will find it odd that deleting an `encodingHistory` item on MVS also deletes it for USS 😋

File: packages/zowe-explorer/src/shared/utils.ts (hunk 8 above)
Comment (KaranP25) [reply packages/zowe-explorer/src/shared/utils.ts:509]: Forgot about that, I have just added one. Thank you.

Comment (rudyflores) [reply packages/zowe-explorer/src/shared/utils.ts:509]: @zFernand0 I believe from what we discussed with @t1m0thyj on this setting is that we wanted them to share this setting between MVS and USS, did this change? I originally intended for these to be independent but it seemed that we wanted to maintain them as shared between each other 🤔

**Review by zFernand0 — APPROVED** [review]

**Comment by zFernand0** [comment]
> zFernand0 I believe from what we discussed with @t1m0thyj on this setting is that we wanted them to share this setting between MVS and USS, did this change? 
> I originally intended for these to be independent but it seemed that we wanted to maintain them as shared between each other 🤔

Thanks for sharing! 🥳 
I partially remember some of those discussions. My apologies for questioning the current implementation.
Since that's what was decided, then I'm onboard 😋 
Thanks for the reminder... 🥳

**Review by rudyflores — APPROVED** [review]
Screenshot 2024-03-21 at 4 51 48 PM

Looks good to me, I did see an issue related to clicking on Encoding History in the drop down and then clicking sessions and the dropdown resizing its size which can be a little odd with all the resizing going on for the user, so we may want to work on keeping the dropdown length consistent in a follow up PR

**Pull request merged**